File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if not intervals:
            return []
        #1. Sort intervals by start.
        intervals.sort(key=lambda x:x[0])

        #2.Initialize a result list merged and push the first interval.
        merged = [intervals[0]]

        #3.For each next interval curr:
        for start,end in intervals[1:]:
            #4.Let last be the last interval in merged.
            last_start,last_end = merged[-1]
            #5.If curr[0] <= last[1], they overlap:
            if start<=last_end:
            #6.Update last[1] = max(last[1], curr[1]).
                merged[-1][1] = max(last_end,end)
            #7.Else (no overlap):
            else:
                #8.Append curr to merged.
                merged.append([start,end])
        return merged

File: 0056-merge-intervals/0056-merge-intervals.go
package main

import (
	"fmt"
	"sort"
)

func merge(intervals [][]int) [][]int {
	if len(intervals) == 0 {
		return [][]int{}
	}
	// 1. Sort by start time
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] < intervals[j][0]
	})
	// 2. Traverse and merge
	merged := [][]int{intervals[0]}
	for i := 1; i < len(intervals); i++ {
		last := merged[len(merged)-1]
		curr := intervals[i]
		// overlap if curr.start <= last.end
		if curr[0] <= last[1] {
			if curr[1] > last[1] {
				last[1] = curr[1]
			}
		} else {
			merged = append(merged, curr)
		}
	}
	return merged
}
